server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/PendingListenersQueueTests.java

```diff
@@ -71,6 +71,45 @@ public void testShouldExecuteOnlyCompleted() throws InterruptedException {
         }
     }
 
+    /**
+     * This test simulates following scenario:
+     *
+     * Master: starts allocation (index 2)
+     * Master: pauses queue
+     * Master: submits new balance calculation (index 2)
+     * Master: allocates using already computed balance (index 1)
+     *
+     * Calculator: starts computing new balance (index 2)
+     * Calculator: new balance is computed (isFresh=true, hasChanges=false), completes the queue for the new balance (index 2)
+     *
+     * Master: completes the queue for applied balance (index 1)
+     * Master: resumes the queue
+     *
+     * Expects: Listener (index 2) should be completed
```

Comment:
In this case index was rolled back to 1 and corresponding listener was stuck until something else advances